### ISSUE TITLE
16x performance difference between sass and node-sass

### ISSUE THREAD

**Issue by dcwarwick** [sass/dart-sass#1534]
We have a jest test for a component library that compiles a stylesheet and compares the result to a snapshot. This test is still using `node-sass`, although we've moved to using `sass` for our build, so I tried to update the test to use `sass` too.

The test looks something like this:
```js
const { renderSync } = require('node-sass');

it("doesn't change the exported CSS for released components", () => {
    expect(
      renderSync({
        file: resolve(__dirname, '../index-without-carbon-released-only.scss'),
        includePaths: [resolve(__dirname, '../../../../node_modules')],
        outputStyle: 'expanded',
      }).css.toString()
    ).toMatchSnapshot();
  });
```

Now this SCSS file loads, indirectly, a big and complex tree of dependent SCSS files, and on my machine it takes about 20 secs to complete the test:
![image](https://user-images.githubusercontent.com/6385315/137381136-5b40a5bc-ecbe-4aec-80f1-ceb24ed26df6.png)

I changed 'node-sass' to 'sass' on the require, as `sass` is already a dependency. I also added `* @jest-environment node` to the header comment. Then I ran the test again, and using `sass` instead of `node-sass` it takes about 345 secs to complete the test:
![image](https://user-images.githubusercontent.com/6385315/137381483-cc05e94c-44bd-4d14-b4de-e92547872dbc.png)

This is a very substantial increase in time, and indeed would cause the total run time of the entire test suite to grow by a factor of more than 3. I am aware that `sass` will not be quite as fast as `node-sass`, but this represents a much larger factor than I was expecting. I'm interested in whether this is what would be expected, and whether there is something about the way I am calling or using `sass` that could be changed to give me better/faster results.

The full test file: https://github.com/carbon-design-system/ibm-cloud-cognitive/blob/main/packages/cloud-cognitive/src/__tests__/styles.test.js

Steps to reproduce the issue:
- Clone the repo https://github.com/carbon-design-system/ibm-cloud-cognitive
- Run `yarn`
- Run `yarn jest styles`. This will run the test using `node-sass`.
- Edit the file /packages/cloud-cognitive/src/__tests__/styles.test.js to add `* @jest-environment node` into the copyright comment at the top of the file and to change 'node-sass' to 'sass' in the first `require`.
- Run `yarn jest styles` again. This will run the test using `sass`.

**Comment by nex3** [comment]
There's also inherently going to be some extra `@import`-resolution overhead for Dart Sass because it supports [import-only files](https://sass-lang.com/documentation/at-rules/import#import-only-files) so it has to check twice as many potential locations for files that might exist.

**Comment by dcwarwick** [comment]
btw, I'm on a Mac. Just out of interest I've loaded an updated test that uses sass instead of node-sass as a PR and run the tests as a github action, here: https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/1340/checks?check_run_id=3911374913. You can see the original (node-sass) "styles.test.js" and the updated (sass) "styles2.test.js" runs here:
![image](https://user-images.githubusercontent.com/6385315/137566312-4c2a4422-2dad-4d00-8a1e-afe2e7464ff0.png)
13s for node-sass, 274s for sass. That is a pretty hefty difference for a simple swap of which API is being called.

**Comment by nex3** [comment]
> Yeah, I could just run the command line compile for my test, but I'm still curious as to why calling through the API is making such a huge difference.

That's a great question. The Dart Sass JS API has some behavior differences from the JS CLI, particularly around the way `@import`s work, to maximize compatibility with the old Node Sass API. Given that there are tens of thousands of imports at play, any differences there will be substantially magnified—that would be my first guess. We're very close to launching the [new JS API](https://sass-lang.com/blog/request-for-comments-new-js-api) for Dart Sass, though, which uses the same import logic as the CLI app. That should bring the JS compilation in line with the JS CLI.

I've packaged up the various performance improvements this issue has inspired into #1535. I'll be re-running our entire benchmark suite to get more solid numbers, but here are the improved numbers for the carbon example:

* LibSass: 6.2s
* Dart Sass compiled to native code: 5.0s (!!)
* Dart Sass compiled to JS: 15.0s

**Comment by nex3** [comment]
@dcwarwick I'd like to add this to our standard benchmark suite, but it looks like it's not yet compatible with Node Sass 6.0.0, and thus not compatible with Node 16. Is there an issue tracking that?

**Comment by dcwarwick** [comment]
Yay! That looks really cool. And in the mean time I will switch to calling out to sass through the CLI rather than in through the JS API to get me some of that 40s action, which is a little slower than node-sass but has the advantage that we'll actually be testing what we're building!

**Comment by dcwarwick** [comment]
What's not compatible? We weren't actually declaring node-sass as a dev dependency, so it was just picking up some random version of node-sass from a dependency, but I just ran the compile with node-sass 6.0.1 and it seemed to work ok.

**Comment by nex3** [comment]
It's just a declaration issue—wherever you're picking up `node-sass` from declares a dependency on `5.x`, which won't compile against Node 16.x. I *think* just changing any 5.x dependencies to 6.x should fix it.

**Comment by dcwarwick** [comment]
Yeah, sorry, the package won't currently work on Node 16.x -- we have a dependency @carbon/bundler which requires node-sass 4.14.1. This dependency is obsolete, and we have an issue to remove it https://github.com/carbon-design-system/ibm-cloud-cognitive/issues/1144 so we might be able to get rid of that in a week or two. I'm running on Node lts/fermium=14 where all's fine.

**Comment by dcwarwick** [comment]
ok, by changing
```js
    expect(
      renderSync({
        file: scssReleasedOnly,
        includePaths,
        outputStyle: 'expanded',
      }).css.toString()
    ).toMatchSnapshot();
```
to
```js
      execSync(
        `sass --style=expanded --load-path ${includePath1} --load-path ${includePath2} ${scssReleasedOnly}`
      ).toString()
    ).toMatchSnapshot();
```


![image](https://user-images.githubusercontent.com/6385315/137599961-63b09fed-3999-4f80-bf38-3a722881a81a.png)

**Comment by dcwarwick** [comment]
@nex3 FWIW we have now removed the obsolete dependency, so there shouldn't be any problem using node 16 with our package now. Relevant tests are now using execFileSync rather than the API and the timings are manageable :-)
